File: python/brayns/core/snapshot/frame_exporter.py
# Copyright (c) 2015-2022 EPFL/Blue Brain Project
# All rights reserved. Do not distribute without permission.
#
# Responsible Author: adrien.fleury@epfl.ch
#
# This file is part of Brayns <https://github.com/BlueBrain/Brayns>
#
# This library is free software; you can redistribute it and/or modify it under
# the terms of the GNU Lesser General Public License version 3.0 as published
# by the Free Software Foundation.
#
# This library is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
# details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this library; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from brayns.network import Instance
from brayns.utils import ImageFormat, Resolution, serialize_view

from ..camera import Camera
from ..renderer import Renderer
from .key_frame import KeyFrame


@dataclass
class FrameExporter:
    """Frame exporter to take multiple snapshots in an optimized way.

    Camera position can be different for each frame using KeyFrame objects.

    Does not modify the current settings of the instance (resolution, camera,
    etc...).

    For parameters with None value, the current ones of the instance are used.

    Exported frames are named using 5 digits filenames (00000.png, 00001.png,
    00003.png, etc...).

    :param frames: Like of key frames to export.
    :type frames: list[KeyFrame]
    :param format: Format of the exported frames, defaults to PNG.
    :type format: ImageFormat, optional
    :param resolution: Resolution of the exported frames, defaults to None.
    :type resolution: Resolution | None, optional
    :param camera: Camera to use for render, defaults to None.
    :type camera: Camera | None, optional
    :param renderer: Renderer to use for render, defaults to None.
    :type renderer: Renderer | None, optional
    :param jpeg_quality: JPEG quality if format is JPEG, defaults to 100.
    :type jpeg_quality: int, optional
    """

    frames: list[KeyFrame]
    format: ImageFormat = ImageFormat.PNG
    resolution: Resolution | None = None
    camera: Camera | None = None
    renderer: Renderer | None = None
    jpeg_quality: int = 100

    def export_frames(self, instance: Instance, folder: str) -> None:
        """Export frames to given folder.

        :param instance: Instance.
        :type instance: Instance
        :param folder: Output folder.
        :type folder: str
        """
        params = _serialize_exporter(self, folder)
        instance.request('export-frames', params)


def _serialize_exporter(exporter: FrameExporter, folder: str) -> dict[str, Any]:
    message: dict[str, Any] = {
        'path': folder,
        'key_frames': [
            _serialize_key_frame(frame)
            for frame in exporter.frames
        ],
        'image_settings': _serialize_image_settings(exporter),
    }
    if exporter.camera is not None:
        message['camera'] = exporter.camera.get_properties_with_name()
    if exporter.renderer is not None:
        message['renderer'] = exporter.renderer.get_properties_with_name()
    return message


def _serialize_key_frame(frame: KeyFrame) -> dict[str, Any]:
    message: dict[str, Any] = {
        'frame_index': frame.index,
    }
    if frame.view is not None:
        message['camera_view'] = serialize_view(frame.view)
    return message


def _serialize_image_settings(exporter: FrameExporter) -> dict[str, Any]:
    message: dict[str, Any] = {
        'format': exporter.format.value,
    }
    if exporter.format is ImageFormat.JPEG:
        message['quality'] = exporter.jpeg_quality
    if exporter.resolution is not None:
        message['size'] = list(exporter.resolution)
    return message

File: brayns/engine/FrameExporter.cpp
/* Copyright 2015-2024 Blue Brain Project/EPFL
 * All rights reserved. Do not distribute without permission.
 * Responsible Author: Nadir Roman <nadir.romanguerrero@epfl.ch>
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License version 3.0 as published
 * by the Free Software Foundation.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include "FrameExporter.h"

#include <brayns/common/Log.h>
#include <brayns/engine/Camera.h>
#include <brayns/engine/Engine.h>
#include <brayns/engine/FrameBuffer.h>
#include <brayns/engine/Renderer.h>
#include <brayns/parameters/ParametersManager.h>
#include <brayns/utils/Filesystem.h>
#include <brayns/utils/image/ImageCodecRegistry.h>
#include <brayns/utils/image/ImageEncoder.h>

namespace brayns
{
namespace
{
void checkExportParameters(const FrameExporter::ExportInfo& input)
{
    if (input.keyFrames.empty())
        throw FrameExportParameterException("No keyframe information");

    if (input.numSamples == 0)
        throw FrameExportParameterException(
            "Number of samples must be greater than 0");

    const auto& path = input.storePath;
    const auto bPath = fs::path(path);
    if (!fs::exists(bPath))
        throw FrameExportParameterException("Frame result path " + path +
                                            " not found");

    // Filesystem allow to query permissions, but no information about
    // current user permissions. To check wether we can write on the target
    // path without using system API, we attempt to create a directory and
    // capture any exception that may be thrown
    std::string testDirBaseName = "test_directory_";
    uint32_t indexTest{0u};
    while (fs::exists(fs::path(testDirBaseName + std::to_string(indexTest))))
        ++indexTest;

    const auto testDirName = testDirBaseName + std::to_string(indexTest);
    const fs::path testDirFilename(testDirName);
    try
    {
        fs::create_directory(bPath / testDirFilename);
    }
    catch (const fs::filesystem_error& fse)
    {
        throw FrameExportParameterException("Cannot write on " + path + ": " +
                                            std::string(fse.what()));
    }
    // Remove test directory on success
    fs::remove(bPath / testDirFilename);

    auto& format = input.imageFormat;
    if (!ImageCodecRegistry::isSupported(format))
    {
        throw FrameExportParameterException("Unknown image file format " +
                                            format);
    }
}
} // namespace

void FrameExporter::startNewExport(ExportInfo&& input)
{
    if (_exportRunning || _exportRequested)
        throw FrameExportInProgressException();

    checkExportParameters(input);

    _currentExport = std::move(input);
    _exportRequested = true;

    brayns::Log::info(
        "Export frame request\n"
        "- Number of frames:\t{}\n"
        "- Samples per pixel:\t{}\n"
        "- Export folder:\t{}",
        _currentExport.keyFrames.size(), _currentExport.numSamples,
        _currentExport.storePath);
}

void FrameExporter::preRender(Camera& camera, Renderer& renderer,
                              FrameBuffer& frameBuffer,
                              ParametersManager& parameters)
{
    if (_exportRequested)
    {
        _exportRunning = true;
        _exportRequested = false;
        _start(camera, renderer, frameBuffer, parameters);
    }

    if (!_exportRunning)
        return;

    // Finish on error
    if (_currentExportError)
    {
        _stop(camera, renderer, parameters);
        return;
    }

    // Export completed
    if (_currentExportKeyFrameIndex >= _currentExport.keyFrames.size())
    {
        _stop(camera, renderer, parameters);
        return;
    }

    // New frame begin, update camera and simulation
    if (_currentExportFrameAccumulation == 0)
    {
        const auto& keyFrame =
            _currentExport.keyFrames[_currentExportKeyFrameIndex];

        camera = keyFrame.camera;
        camera.setProperties(keyFrame.cameraParameters);
        parameters.getAnimationParameters().setFrame(keyFrame.frameIndex);
    }
}

void FrameExporter::postRender(FrameBuffer& frameBuffer)
{
    if (!_exportRunning)
        return;

    ++_currentExportFrameAccumulation;
    // Frame completed, write to disk
    if (_currentExportFrameAccumulation >= _currentExport.numSamples)
    {
        auto frameNumberName = _currentExportKeyFrameIndex;
        if (_currentExport.nameImageAfterSimulationFrameIndex)
            frameNumberName =
                _currentExport.keyFrames[_currentExportKeyFrameIndex]
                    .frameIndex;
        _writeImageToDisk(frameBuffer, frameNumberName);

        _currentExportFrameAccumulation = 0;
        ++_currentExportKeyFrameIndex;
    }
}

double FrameExporter::getExportProgress()
{
    if (!_exportRunning)
        throw FrameExportNotRunningException();

    if (_currentExportError)
        throw std::runtime_error(_currentExportErrorMessage);

    const auto numKeyframes = _currentExport.keyFrames.size();
    const auto spp = _currentExport.numSamples;

    const auto totalFrameCount = numKeyframes * spp;
    const auto progress =
        _currentExportKeyFrameIndex * spp + _currentExportFrameAccumulation;

    return static_cast<double>(progress) / static_cast<double>(totalFrameCount);
}

void FrameExporter::_saveState(Camera& camera, ParametersManager& parameters)
{
    _originalState = {};

    const auto& renderParams = parameters.getRenderingParameters();
    _originalState.hasAccumulation = renderParams.getAccumulation();
    _originalState.accumulationSize = renderParams.getMaxAccumFrames();

    _originalState.camera = camera;
    _originalState.cameraProperties = camera.getPropertyMap();
}

void FrameExporter::_restoreState(Camera& camera, ParametersManager& parameters)
{
    auto& renderParams = parameters.getRenderingParameters();
    renderParams.setAccumulation(_originalState.hasAccumulation);
    renderParams.setMaxAccumFrames(_originalState.accumulationSize);

    camera = _originalState.camera;
    camera.setProperties(_originalState.cameraProperties);

    camera.markModified();
    camera.commit();
}

void FrameExporter::_start(Camera& camera, Renderer& renderer,
                           FrameBuffer& frameBuffer,
                           ParametersManager& parameters) noexcept
{
    // Initialize export status
    _exportRunning = true;
    _currentExportKeyFrameIndex = 0u;
    _currentExportFrameAccumulation = 0u;
    _currentExportError = 0;
    _currentExportErrorMessage = "";

    _saveState(camera, parameters);

    auto& renderParams = parameters.getRenderingParameters();
    renderParams.setMaxAccumFrames(_currentExport.numSamples + 1);

    renderer.markModified();
    renderer.commit();

    frameBuffer.clear();
}

void FrameExporter::_stop(Camera& camera, Renderer& renderer,
                          ParametersManager& parameters) noexcept
{
    _exportRunning = false;
    _currentExportKeyFrameIndex = 0u;
    _currentExportFrameAccumulation = 0u;

    _restoreState(camera, parameters);

    renderer.markModified();
    renderer.commit();
}

void FrameExporter::_writeImageToDisk(FrameBuffer& frameBuffer,
                                      const uint32_t frameNumberName)
{
    auto image = frameBuffer.getImage();

    char frame[7];
    sprintf(frame, "%05d", static_cast<int32_t>(frameNumberName));
    auto& basePath = _currentExport.storePath;
    auto& extension = _currentExport.imageFormat;
    auto filename = basePath + '/' + frame + "." + extension;
    auto quality = _currentExport.imageQuality;

    try
    {
        ImageEncoder::save(image, filename, quality);
    }
    catch (const std::runtime_error& e)
    {
        _currentExportError = true;
        _currentExportErrorMessage = e.what();
        brayns::Log::error("{}", _currentExportErrorMessage);
        return;
    }

    frameBuffer.clear();

    brayns::Log::info("Frame saved to {}", filename);
}
} // namespace brayns
